### PR TITLE
Single result queries

### PR DESCRIPTION
**Pull request by Blackbaud-EricSlater**
This is a port of this PR from the old repository...
https://github.com/microsoft/spring-data-cosmosdb/pull/480

It now includes reactive support as requested in the other PR.

### CODE REVIEW

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - thank you for creating this PR, as I informed you earlier, the repo is currently being migrated to new azure-cosmos V4 dependency SDK. 

This is the PR I have currently open for that - https://github.com/Azure/azure-sdk-for-java/pull/13229

However, your timing couldn't be better, I plan to merge that PR today - so once it is in master today, can you please rebase your PR with master and update it accordingly and we can try to get your PR in this week, if not - definitely by early next week.

**Comment by Blackbaud-EricSlater** [comment]
sounds good @kushagraThapar thanks!

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater my PR is in now - feel free to rebase your PR to latest master and we can review this.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - Can you access the failed builds ?
It looks like the code is not compiling. 
One of the issues I found out is the use of `CosmosDBAccessException` in your PR. 

In master, the exception is called - `CosmosAccessException` 
As part of rebranding of CosmosDb to Cosmos, I have updated most of the code in master to reflect this rebranding. 
So I have renamed most of the classes from `CosmosDb*` to `Cosmos*`

Please update your code.

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar yep Ill have a fix up shortly for the compile issues and the other style issues that are failing the build. Sorry i should have pushed this all up at once instead of just pushing up the rebase.

**Comment by kushagraThapar** [comment]
> @kushagraThapar yep Ill have a fix up shortly for the compile issues and the other style issues that are failing the build. Sorry i should have pushed this all up at once instead of just pushing up the rebase.

No worries at all :) 
Since we recently brought spring-cosmos to azure-sdk-for-java repo, there have been some new changes on the code styling, checkstyles, spotbugs, and java docs styles. 
We all are learning together :) 

I just wanted to make sure you can access the builds and see the errors so it is much easier for you to fix them. 

Let me know if you have any more questions / concerns regarding these issues.

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java [hunk 1]
```diff
@@ -54,6 +56,34 @@ public Object execute(DocumentQuery query, Class<?> type, String container) {
...
+
+        public SingleEntityExecution(ReactiveCosmosOperations operations, ReturnedType returnedType) {
+            this.operations = operations;
+            this.returnedType = returnedType;
+        }
+
+        @Override
+        public Object execute(DocumentQuery query, Class<?> type, String container) {
+            return operations.find(query, type, container)
+                .buffer(2)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java:75]: why do we need this `buffer(2)` ?

File: sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveCourseRepository.java [hunk 2]
```diff
@@ -34,4 +35,12 @@
      * @return Course list
      */
     Flux<Course> findByNameOrDepartmentAllIgnoreCase(String name, String department);
+
+    /**
+     * Find a single Course list by name
+     * @param name name
+     * @return Course list
+     */
+    Mono<Course> findByName(String name);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveCourseRepository.java:44]: Shouldn't this also be `findOneByName` ?
I am still unclear on what if there are multiple items by `name` ?

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveCourseRepository.java:44]: yea i will rename that for clarity. But i do want to point out that the determination of the execution path is based on the return type. We are not using the PartTree class to parse the method name to determine the users intent so it behaves the same regardless of whether the user specifies "findOne". The fact that the method is returning a Mono is what makes it expect one item.

In the event that there are multiple items returned, we will throw a CosmosAccessException with an error message saying "Too many results - Expected Mono<Course> but query returned multiple results".

This is consistent with the non reactive path and is demonstrated in the test ReactiveCourseRepositoryIT.testFindOneShouldFailIfMultipleResultsReturned

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java (hunk 1 above)
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java:75]: We want to ensure only a single result is returned by the query. However, the results are returned as a stream. This allows us to check that the stream only contains a single item without retrieving all of the streams contents.

File: sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveCourseRepository.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveCourseRepository.java:44]: Sounds great, thank you for the explanation.

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/query/ReactiveCosmosQueryExecution.java:75]: Sounds good!

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - Because of the new `if statement` added in your PR, it is causing the Java 11 code coverage to fail. Can you please come up with some unit tests to test the single query / multi query execution code path ?
I am thinking if you can add a unit test case for `AbstractCosmosQuery` and `AbstractReactiveCosmosQuery` - just something to start with, and then I can add more test cases for these going forward.

**Comment by Blackbaud-EricSlater** [comment]
added tests AbstractCosmosQuery and AbstractReactiveCosmosQuery. This required slight changes to these classes in order to wire them up properly. 

I tried initially to use PartTreeCosmosQuery and ReactivePartTreeCosmosQuery instead of adding concrete test implementations of those classes but they had dependency graphs that could not be satisfied in a unit test. They depended on classes that had package protected constructors that i couldn't access or were calling methods that were final and couldn't be mocked.

I also had to make getExecution(...) a protected method so that it was callable from a test. 

I also had to wrap method.isPageQuery() in a protected method that i could override because it is final and couldnt be mocked.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - the new changes regarding unit test look good. Thank you so much for making this feature compliant with our existing standards, great work!

I will get this in as soon as the CI passes.

**Pull request merged**